README.md
# KQ-Pad

- Current Version: **v0.1**

Notepad knockoff made with Python and QT (PySide2).

# Compile or Run from Source

## Ubuntu/Debian

1. Run the commands below (line by line) to install the dependencies

```bash
sudo apt update && sudo apt upgrade
sudo apt install python3 python3-pip git -y
pip3 install -U PyInstaller PySide2
```

- Run the commands below to compile the software (SLOW + Resource Heavy)

```bash
cd "~/Desktop/"
mkdir temp && cd temp
git clone https://github.com/hirusha-adi/Notes-QT.git
cd Notes-QT

pyinstaller notes.py --onefile --noconfirm
# or
python3 -m PyInstaller notes.py --onefile --noconfirm

cd ./dist
sudo cp notes /usr/bin/notes
sudo cp notes.png /usr/bin/notes.png
cd ~
rm -rf "~/Desktop/temp/"
```

- Run from source (Fast + Lightweight)

```bash
cd "~/Desktop/"
mkdir temp && cd temp
git clone https://github.com/hirusha-adi/Notes-QT.git
cd Notes-QT
sudo mv notes.py /usr/bin/notes
sudo cp notes.png /usr/bin/notes.png
cd ~
rm -rf "~/Desktop/temp/"
```

## Arch Linux

1. Run the commands below (line by line) to install the dependencies

```bash
sudo pacman -Syyuu python python-pip git --noconfirm
pip install PySide2 PyInstaller
```

- Run the commands below to compile the software (SLOW + Resource Heavy)

```bash
cd "~/Desktop/"
mkdir temp && cd temp
git clone https://github.com/hirusha-adi/Notes-QT.git
cd Notes-QT

pyinstaller notes.py --onefile --noconfirm
# or
python3 -m PyInstaller notes.py --onefile --noconfirm

cd ./dist
sudo cp notes /usr/bin/notes
sudo cp notes.png /usr/bin/notes.png
cd ~
rm -rf "~/Desktop/temp/"
```

- Run from source (Fast + Lightweight)

```bash
cd "~/Desktop/"
mkdir temp && cd temp
git clone https://github.com/hirusha-adi/Notes-QT.git
cd Notes-QT
sudo mv notes.py /usr/bin/notes
sudo cp notes.png /usr/bin/notes.png
cd ~
rm -rf "~/Desktop/temp/"
```

## Microsoft Windows

1. Download and install Python3. Make sure to check "Add to PATH" on the first screen of installation
2. Open `cmd` and run the command below to install the requirements

```bash
pip install -U PyInstaller PySide2
```

3. Clone the repository and `cd` into it OR download the source code as `.zip` or `.tar.gz` and extract it. Open a `cmd` in the extracted folder.
4. Run the commands below to to compile

```bash
pyinstaller --noconfirm --onefile notes.py
```

5. Open the `.\dist` folder and you will find your compiled executable file (`.\dist\notes.exe`)
6. Copy the `notes.exe` to the `%system32%` folder with the `notes.png` (this file is optional) or add the current folder to PATH
7. Run `notes` in the `cmd` or `RUN` (`Win+R`) to start the program

# Credits

- Images

  1. Logo taken from [Wikimedia](https://upload.wikimedia.org/wikipedia/commons/thumb/7/71/Notepad_icon.svg/768px-Notepad_icon.svg.png)

- Sending me images of Notepad
  1. Discord user `OLIVER`
  2. Discord user `OkSheBroken`
  3. Discord user `key`

# Screenshots (on Arch Linux with KDE Breeze Dark)

![image](https://user-images.githubusercontent.com/36286877/149811059-71a4d02d-cee7-4145-ae10-a1be8c0d519b.png)

<br>

- Change Font
<br>

![image](https://user-images.githubusercontent.com/36286877/149810619-1b4782b9-9c36-437a-a74d-58958f9f80f3.png)

- Print
<br>

![image](https://user-images.githubusercontent.com/36286877/149810774-8e3759ff-8933-49e7-8e69-3ec3e157c699.png)

- Print Preview
<br>

![image](https://user-images.githubusercontent.com/36286877/149810906-439d0591-c7d3-40e3-99ed-cfded510d54a.png)

- Save As
<br>

![image](https://user-images.githubusercontent.com/36286877/149810966-29d2522d-8713-4b61-8638-e67d80389be8.png)



changelog.md
# v0.1

- 2022/01/17

  - word wrap (on and off) - defaults to no wrap
  - added save and save as, works as intended
  - added new file function (unstable, barely works, pls i help i beg)
  - added print and print preview
  - added undo and redo
  - added ability to change the font related stuff
  - added zoom in and zoom out
  - added help, about and send feedback

- 2022/01/16
  - Created GUI
  - Added all menu's and keyboard shortcuts
  - open file


test.py
import sys
from PySide2.QtWidgets import QApplication, QFileDialog


def get_filename():
    app = QApplication([])
    options = QFileDialog.Options()
    options |= QFileDialog.DontUseNativeDialog
    fileName, _ = QFileDialog.getOpenFileName(
        None,
        "QFileDialog.getOpenFileName()",
        "",
        "All Files (*);;Text Files (*.txt)",
        options=options,
    )
    return fileName


if __name__ == "__main__":
    filename = get_filename()
    if filename:
        print(filename)


main-bak.py
# import sys
# from PySide2 import QtCore, QtGui, QtWidgets
# from PySide2.QtUiTools import QUiLoader

# loader = QUiLoader()
# app = QtWidgets.QApplication(sys.argv)
# window = loader.load("Notes.ui", None)
# window.show()
# app.exec_()

from PySide2.QtWidgets import QApplication, QMainWindow, QAction, QTextEdit
import sys
from PySide2.QtGui import QIcon, QFont


class Window(QMainWindow):
    def __init__(self):
        super().__init__()

        self.setWindowTitle("Simple Notepad Application")
        self.setGeometry(300, 200, 1280, 600)

        self.textEdit = QTextEdit(self)
        self.textEdit.setFont(QFont('Sanserif', 13))
        self.setCentralWidget(self.textEdit)

        self.setIcon()

        self.create_menu()
        self.show()

    def setIcon(self):
        appIcon = QIcon("icon.png")
        self.setWindowIcon(appIcon)

    def create_menu(self):
        mainMenu = self.menuBar()
        fileMenu = mainMenu.addMenu('File')
        viewMenu = mainMenu.addMenu('View')
        editMenu = mainMenu.addMenu('Edit')
        fontMenu = mainMenu.addMenu('Font')
        helpMenu = mainMenu.addMenu('Help')

        # Main Menu: File
        # --------------------------------------------------------
        openAction = QAction(QIcon('open.png'), "Open", self)
        openAction.setShortcut('Ctrl+N')

        openAction = QAction(QIcon('open.png'), "Open", self)
        openAction.setShortcut('Ctrl+O')

        saveAction = QAction(QIcon('save.png'), "Save", self)
        saveAction.setShortcut('Ctrl+S')

        exitAction = QAction(QIcon('exit.png'), "Exit", self)
        exitAction.setShortcut('Ctrl+X')
        exitAction.triggered.connect(self.exit_app)

        fileMenu.addAction(openAction)
        fileMenu.addAction(saveAction)
        fileMenu.addAction(exitAction)

    def exit_app(self):
        self.close()


myapp = QApplication(sys.argv)
window = Window()


myapp.exec_()
sys.exit()


original.py
# # -*- coding: utf-8 -*-

# ################################################################################
# # Form generated from reading UI file 'NotesnPYfYl.ui'
# ##
# # Created by: Qt User Interface Compiler version 5.15.2
# ##
# # WARNING! All changes made in this file will be lost when recompiling UI file!
# ################################################################################
# import sys

from PySide2.QtCore import *
from PySide2.QtGui import *
from PySide2.QtWidgets import *


class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        if not MainWindow.objectName():
            MainWindow.setObjectName(u"MainWindow")
        MainWindow.resize(822, 600)

        self.actionNew = QAction(MainWindow)
        self.actionNew.setObjectName(u"actionNew")
        self.actionOpen = QAction(MainWindow)
        self.actionOpen.setObjectName(u"actionOpen")
        self.actionSave = QAction(MainWindow)
        self.actionSave.setObjectName(u"actionSave")
        self.actionSave_As = QAction(MainWindow)
        self.actionSave_As.setObjectName(u"actionSave_As")
        self.actionPage_Setup = QAction(MainWindow)
        self.actionPage_Setup.setObjectName(u"actionPage_Setup")
        self.actionPage_Setup_2 = QAction(MainWindow)
        self.actionPage_Setup_2.setObjectName(u"actionPage_Setup_2")
        self.actionPrint = QAction(MainWindow)
        self.actionPrint.setObjectName(u"actionPrint")
        self.actionExit = QAction(MainWindow)
        self.actionExit.setObjectName(u"actionExit")
        self.actionUndo = QAction(MainWindow)
        self.actionUndo.setObjectName(u"actionUndo")
        self.actionCut = QAction(MainWindow)
        self.actionCut.setObjectName(u"actionCut")
        self.actionCopy = QAction(MainWindow)
        self.actionCopy.setObjectName(u"actionCopy")
        self.actionPaste = QAction(MainWindow)
        self.actionPaste.setObjectName(u"actionPaste")
        self.actionDelete = QAction(MainWindow)
        self.actionDelete.setObjectName(u"actionDelete")
        self.actionFind = QAction(MainWindow)
        self.actionFind.setObjectName(u"actionFind")
        self.actionFind_Next = QAction(MainWindow)
        self.actionFind_Next.setObjectName(u"actionFind_Next")
        self.actionReplace = QAction(MainWindow)
        self.actionReplace.setObjectName(u"actionReplace")
        self.actionGo_To = QAction(MainWindow)
        self.actionGo_To.setObjectName(u"actionGo_To")
        self.actionSelect_All = QAction(MainWindow)
        self.actionSelect_All.setObjectName(u"actionSelect_All")
        self.actionTime_Date = QAction(MainWindow)
        self.actionTime_Date.setObjectName(u"actionTime_Date")
        self.actionWord_Wrap = QAction(MainWindow)
        self.actionWord_Wrap.setObjectName(u"actionWord_Wrap")
        self.actionFont = QAction(MainWindow)
        self.actionFont.setObjectName(u"actionFont")
        self.actionZoom = QAction(MainWindow)
        self.actionZoom.setObjectName(u"actionZoom")
        self.actionStatus_Bar = QAction(MainWindow)
        self.actionStatus_Bar.setObjectName(u"actionStatus_Bar")
        self.actionView_Help = QAction(MainWindow)
        self.actionView_Help.setObjectName(u"actionView_Help")
        self.actionAbout_Notes = QAction(MainWindow)
        self.actionAbout_Notes.setObjectName(u"actionAbout_Notes")
        self.centralwidget = QWidget(MainWindow)
        self.centralwidget.setObjectName(u"centralwidget")
        self.plainTextEdit = QPlainTextEdit(self.centralwidget)
        self.plainTextEdit.setObjectName(u"plainTextEdit")
        self.plainTextEdit.setGeometry(QRect(0, 0, 821, 551))
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QMenuBar(MainWindow)
        self.menubar.setObjectName(u"menubar")
        self.menubar.setGeometry(QRect(0, 0, 822, 29))
        self.menuLogin = QMenu(self.menubar)
        self.menuLogin.setObjectName(u"menuLogin")
        self.menuEdit = QMenu(self.menubar)
        self.menuEdit.setObjectName(u"menuEdit")
        self.menuFormat = QMenu(self.menubar)
        self.menuFormat.setObjectName(u"menuFormat")
        self.menuView = QMenu(self.menubar)
        self.menuView.setObjectName(u"menuView")
        self.menuHelp = QMenu(self.menubar)
        self.menuHelp.setObjectName(u"menuHelp")
        MainWindow.setMenuBar(self.menubar)
        self.statusbar = QStatusBar(MainWindow)
        self.statusbar.setObjectName(u"statusbar")
        MainWindow.setStatusBar(self.statusbar)

        self.menubar.addAction(self.menuLogin.menuAction())
        self.menubar.addAction(self.menuEdit.menuAction())
        self.menubar.addAction(self.menuFormat.menuAction())
        self.menubar.addAction(self.menuView.menuAction())
        self.menubar.addAction(self.menuHelp.menuAction())
        self.menuLogin.addAction(self.actionNew)
        self.menuLogin.addAction(self.actionOpen)
        self.menuLogin.addAction(self.actionSave)
        self.menuLogin.addAction(self.actionSave_As)
        self.menuLogin.addSeparator()
        self.menuLogin.addAction(self.actionPage_Setup_2)
        self.menuLogin.addAction(self.actionPrint)
        self.menuLogin.addSeparator()
        self.menuLogin.addAction(self.actionExit)
        self.menuEdit.addAction(self.actionUndo)
        self.menuEdit.addSeparator()
        self.menuEdit.addAction(self.actionCut)
        self.menuEdit.addAction(self.actionCopy)
        self.menuEdit.addAction(self.actionPaste)
        self.menuEdit.addAction(self.actionDelete)
        self.menuEdit.addSeparator()
        self.menuEdit.addAction(self.actionFind)
        self.menuEdit.addAction(self.actionFind_Next)
        self.menuEdit.addAction(self.actionReplace)
        self.menuEdit.addAction(self.actionGo_To)
        self.menuEdit.addSeparator()
        self.menuEdit.addAction(self.actionSelect_All)
        self.menuEdit.addAction(self.actionTime_Date)
        self.menuFormat.addAction(self.actionWord_Wrap)
        self.menuFormat.addAction(self.actionFont)
        self.menuView.addAction(self.actionZoom)
        self.menuView.addAction(self.actionStatus_Bar)
        self.menuHelp.addAction(self.actionView_Help)
        self.menuHelp.addAction(self.actionAbout_Notes)

        self.retranslateUi(MainWindow)

        QMetaObject.connectSlotsByName(MainWindow)
    # setupUi

    def retranslateUi(self, MainWindow):
        MainWindow.setWindowTitle(
            QCoreApplication.translate("MainWindow", u"Notes", None))
        self.actionNew.setText(
            QCoreApplication.translate("MainWindow", u"New", None))
        self.actionOpen.setText(
            QCoreApplication.translate("MainWindow", u"Open", None))
        self.actionSave.setText(
            QCoreApplication.translate("MainWindow", u"Save", None))
        self.actionSave_As.setText(
            QCoreApplication.translate("MainWindow", u"Save As", None))
        self.actionPage_Setup.setText(
            QCoreApplication.translate("MainWindow", u"Page Setup", None))
        self.actionPage_Setup_2.setText(
            QCoreApplication.translate("MainWindow", u"Page Setup", None))
        self.actionPrint.setText(
            QCoreApplication.translate("MainWindow", u"Print", None))
        self.actionExit.setText(
            QCoreApplication.translate("MainWindow", u"Exit", None))
        self.actionUndo.setText(
            QCoreApplication.translate("MainWindow", u"Undo", None))
        self.actionCut.setText(
            QCoreApplication.translate("MainWindow", u"Cut", None))
        self.actionCopy.setText(
            QCoreApplication.translate("MainWindow", u"Copy", None))
        self.actionPaste.setText(
            QCoreApplication.translate("MainWindow", u"Paste", None))
        self.actionDelete.setText(
            QCoreApplication.translate("MainWindow", u"Delete", None))
        self.actionFind.setText(
            QCoreApplication.translate("MainWindow", u"Find", None))
        self.actionFind_Next.setText(
            QCoreApplication.translate("MainWindow", u"Find Next", None))
        self.actionReplace.setText(
            QCoreApplication.translate("MainWindow", u"Replace", None))
        self.actionGo_To.setText(
            QCoreApplication.translate("MainWindow", u"Go To", None))
        self.actionSelect_All.setText(
            QCoreApplication.translate("MainWindow", u"Select All", None))
        self.actionTime_Date.setText(
            QCoreApplication.translate("MainWindow", u"Time/Date", None))
        self.actionWord_Wrap.setText(
            QCoreApplication.translate("MainWindow", u"Word Wrap", None))
        self.actionFont.setText(QCoreApplication.translate(
            "MainWindow", u"Font...", None))
        self.actionZoom.setText(
            QCoreApplication.translate("MainWindow", u"Zoom", None))
        self.actionStatus_Bar.setText(
            QCoreApplication.translate("MainWindow", u"Status Bar", None))
        self.actionView_Help.setText(
            QCoreApplication.translate("MainWindow", u"View Help", None))
        self.actionAbout_Notes.setText(
            QCoreApplication.translate("MainWindow", u"About Notes", None))
        self.menuLogin.setTitle(
            QCoreApplication.translate("MainWindow", u"File", None))
        self.menuEdit.setTitle(
            QCoreApplication.translate("MainWindow", u"Edit", None))
        self.menuFormat.setTitle(
            QCoreApplication.translate("MainWindow", u"Format", None))
        self.menuView.setTitle(
            QCoreApplication.translate("MainWindow", u"View", None))
        self.menuHelp.setTitle(
            QCoreApplication.translate("MainWindow", u"Help", None))
    # retranslateUi


class MainWindow(QMainWindow, Ui_MainWindow):
    def __init__(self):
        Ui_MainWindow.__init__(self)
        QMainWindow.__init__(self)

        # Initialize UI
        self.setupUi(self)

    def tr(self, text):
        return QObject.tr(self, text)

        # import sys
        # from PySide2 import QtCore, QtGui, QtWidgets
        # from PySide2.QtUiTools import QUiLoader

        # loader = QUiLoader()
        # app = QtWidgets.QApplication(sys.argv)
        # window = loader.load("Notes.ui", None)
        # window.show()
        # app.exec_()


if __name__ == "__main__":
    import sys
    app = QApplication(sys.argv)
    main_window = MainWindow()
    main_window.show()
    sys.exit(app.exec_())


main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import webbrowser
from datetime import datetime

import clipboard  # pip install clipboard
from PySide2.QtCore import *
from PySide2.QtGui import *
from PySide2.QtPrintSupport import *
from PySide2.QtWidgets import *


class Window(QMainWindow):
    def __init__(self):
        super().__init__()

        # main stuff
        self.NAME = "KQ-Pad"
        self.VERSION = "v0.1"
        self.TITLE = self.NAME + " | " + self.VERSION
        self.HELP_LINK = r"https://github.com/hirusha-adi/Notes-QT/blob/main/HELP.md"

        # Needed for the functions to function
        self.fileName = None
        self.wordWrapMode = True
        self.currentZoom = 1

        # Code
        self.setWindowTitle(self.TITLE)
        self.setGeometry(300, 200, 650, 500)

        self.textEdit = QTextEdit(self)
        self.textEdit.setFont(QFont('Sanserif', 13))
        self.setCentralWidget(self.textEdit)
        self._setWordWrapFalse()
        if not(self.textEdit.isUndoRedoEnabled()):
            self.textEdit.setUndoRedoEnabled(enable=True)

        self.setIcon()
        self.create_menu()
        self.show()

    def setIcon(self):
        appIcon = QIcon("notes.png")
        self.setWindowIcon(appIcon)

    def create_menu(self):
        mainMenu = self.menuBar()
        fileMenu = mainMenu.addMenu('File')
        editMenu = mainMenu.addMenu('Edit')
        fontMenu = mainMenu.addMenu('Format')
        viewMenu = mainMenu.addMenu('View')
        helpMenu = mainMenu.addMenu('Help')

        # Main Menu: File
        # --------------------------------------------------------
        newFileAction = QAction("New", self)
        newFileAction.setShortcut('Ctrl+N')
        newFileAction.triggered.connect(self.newFileAction_Clicked)

        openAction = QAction("Open", self)
        openAction.setShortcut('Ctrl+O')
        openAction.triggered.connect(self.openAction_Clicked)

        saveAction = QAction("Save", self)
        saveAction.setShortcut('Ctrl+S')
        saveAction.triggered.connect(self.saveAction_Clicked)

        saveAsAction = QAction("Save As", self)
        saveAsAction.setShortcut('Ctrl+Shift+S')
        saveAsAction.triggered.connect(self.saveAsAction_Clicked)

        printAction = QAction('Print', self)
        printAction.setShortcut('Ctrl+P')
        printAction.triggered.connect(self.printAction_Clicked)

        printPreviewAction = QAction("Print Preview", self)
        printPreviewAction.triggered.connect(self.printPreviewAction_Clicked)

        exitAction = QAction("Exit", self)
        exitAction.setShortcut('Ctrl+X')
        exitAction.triggered.connect(self.exit_app)

        # Main Menu: Edit
        # --------------------------------------------------------
        undoAction = QAction("Undo", self)
        undoAction.setShortcut('Ctrl+Z')
        undoAction.triggered.connect(self.undoAction_Clicked)

        redoAction = QAction("Redo", self)
        redoAction.setShortcut('Ctrl+Y')
        redoAction.triggered.connect(self.redoAction_Clicked)

        cutAction = QAction("Cut", self)
        cutAction.setShortcut('Ctrl+X')
        cutAction.triggered.connect(self.cutAction_Clicked)

        copyAction = QAction("Copy", self)
        copyAction.setShortcut('Ctrl+C')
        copyAction.triggered.connect(self.copyAction_Clicked)

        pasteAction = QAction("Paste", self)
        pasteAction.setShortcut('Ctrl+V')
        pasteAction.triggered.connect(self.pasteAction_Clicked)

        deleteAction = QAction("Delete", self)
        deleteAction.setShortcut('Del')
        deleteAction.triggered.connect(self.deleteAction_Clicked)
        deleteAction.setEnabled(False)

        findAction = QAction("Find", self)
        findAction.setShortcut('Ctrl+F')
        findAction.triggered.connect(self.findAction_Clicked)
        findAction.setEnabled(False)

        findNextAction = QAction("Find Next", self)
        findNextAction.setShortcut('F3')
        findNextAction.triggered.connect(self.findNextAction_Clicked)
        findNextAction.setEnabled(False)

        replaceAction = QAction("Replace", self)
        replaceAction.setShortcut('Ctrl+H')
        replaceAction.triggered.connect(self.replaceAction_Clicked)
        replaceAction.setEnabled(False)

        goToAction = QAction("Go To", self)
        goToAction.setShortcut('Ctrl+G')
        goToAction.triggered.connect(self.goToAction_Clicked)
        goToAction.setEnabled(False)

        selectAllAction = QAction("Select All", self)
        selectAllAction.setShortcut('Ctrl+A')
        selectAllAction.triggered.connect(self.selectAllAction_Clicked)

        timeDateAction = QAction("Time/Date", self)
        timeDateAction.setShortcut('F5')
        timeDateAction.triggered.connect(self.timeDateAction_Clicked)

        # Main Menu: Format
        # --------------------------------------------------------
        wordWrapAction = QAction("Word Wrap", self)
        wordWrapAction.setShortcut('Ctrl+.')
        wordWrapAction.triggered.connect(self.wordWrapAction_Clicked)

        fontAction = QAction("Font...", self)
        fontAction.triggered.connect(self.fontAction_Clicked)

        # Main Menu: View
        # --------------------------------------------------------
        zoomInAction = QAction("Zoom In", self)
        zoomInAction.setShortcut('Ctrl+=')
        zoomInAction.triggered.connect(self.zoomInAction_Clicked)

        zoomOutAction = QAction("Zoom Out", self)
        zoomOutAction.setShortcut('Ctrl+-')
        zoomOutAction.triggered.connect(self.zoomOutAction_Clicked)

        restoreDefaultZoomAction = QAction("Restore Default Zoom", self)
        restoreDefaultZoomAction.triggered.connect(
            self.restoreDefaultZoomAction_Clicked)
        restoreDefaultZoomAction.setEnabled(False)

        # Main Menu: Help
        # --------------------------------------------------------
        viewHelpAction = QAction("View Help", self)
        viewHelpAction.triggered.connect(self.zoomOutAction_Clicked)

        sendFeedbackAction = QAction("Send Feedback", self)
        sendFeedbackAction.triggered.connect(self.sendFeedbackAction_Clicked)

        aboutAction = QAction("About " + self.NAME, self)
        aboutAction.triggered.connect(self.aboutAction_Clicked)

        # FINALLY, Adding everything
        # --------------------------------------------------------
        # File Menu
        fileMenu.addAction(newFileAction)  # new file
        fileMenu.addAction(openAction)  # open
        fileMenu.addAction(saveAction)  # save
        fileMenu.addAction(saveAsAction)  # save as
        fileMenu.addSeparator()  # seperator
        fileMenu.addAction(printAction)  # print
        fileMenu.addAction(printPreviewAction)  # print preview
        fileMenu.addSeparator()  # seperator
        fileMenu.addAction(exitAction)  # exit

        # Edit Menu
        editMenu.addAction(undoAction)  # undo
        editMenu.addAction(redoAction)  # redo
        editMenu.addSeparator()  # seperator
        editMenu.addAction(cutAction)  # cut
        editMenu.addAction(copyAction)  # copy
        editMenu.addAction(pasteAction)  # paste
        # editMenu.addAction(deleteAction)  # delete
        editMenu.addSeparator()  # seperator
        editMenu.addAction(findAction)  # find
        editMenu.addAction(findNextAction)  # find next
        editMenu.addAction(replaceAction)  # find
        editMenu.addAction(goToAction)  # find
        editMenu.addSeparator()  # seperator
        editMenu.addAction(selectAllAction)  # find
        editMenu.addAction(timeDateAction)  # find

        # Format Menu
        fontMenu.addAction(wordWrapAction)
        fontMenu.addAction(fontAction)

        # View Menu
        viewMenu.addAction(zoomInAction)  # zoom in
        viewMenu.addAction(zoomOutAction)  # zoom out
        # restore to default zoom level
        viewMenu.addAction(restoreDefaultZoomAction)
        viewMenu.addSeparator()  # seperator

        # Help Menu
        helpMenu.addAction(viewHelpAction)  # view help
        helpMenu.addAction(sendFeedbackAction)  # send feedback
        helpMenu.addSeparator()  # seperator
        helpMenu.addAction(aboutAction)  # about

    def _getFilenameToOpen(self):
        options = QFileDialog.Options()
        options |= QFileDialog.DontUseNativeDialog
        fileName, _ = QFileDialog.getOpenFileName(
            None,
            "Open File | " + self.TITLE,
            "",
            "All Files (*);;Text Files (*.txt)",
            options=options,
        )
        return fileName

    def openAction_Clicked(self):
        # https://stackoverflow.com/questions/60977801/file-browser-with-pyside2-get-the-path-of-the-file-and-then-kill-the-gui

        if self.fileName:
            with open(file=self.fileName, mode="rt", encoding="utf-8") as fileOpen:
                fileOpenContent = fileOpen.read()
            if self._getTextBoxContent() != fileOpenContent:
                self._newFileCheckSavedOrNot()

        self.fileName = self._getFilenameToOpen()
        if self.fileName:
            with open(file=self.fileName, mode="rt", encoding="utf-8") as fileOpen:
                fileOpenContent = fileOpen.read()
            self.textEdit.setPlainText(fileOpenContent)

    def _setWordWrapFalse(self):
        self.textEdit.setLineWrapMode(QTextEdit.NoWrap)

    def _setWordWrapTrue(self):
        self.textEdit.setLineWrapMode(QTextEdit.WidgetWidth)

    def wordWrapAction_Clicked(self):
        if self.wordWrapMode == False:
            self._setWordWrapFalse()
            self.wordWrapMode = True

        elif self.wordWrapMode == True:
            self._setWordWrapTrue()
            self.wordWrapMode = False

    def _getSaveAsFileName(self):
        # https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QFileDialog.html#PySide2.QtWidgets.PySide2.QtWidgets.QFileDialog.getSaveFileName
        options = QFileDialog.Options()
        options |= QFileDialog.DontUseNativeDialog
        fileName, _ = QFileDialog.getSaveFileName(
            self,
            "Save file | " + self.TITLE,
            "",
            "All Files(*);;Text Files (*.txt)")
        return fileName

    def _getTextBoxContent(self):
        return str(self.textEdit.toPlainText())

    def saveAsAction_Clicked(self):
        fileNameSave = self._getSaveAsFileName()
        if fileNameSave:
            self.fileName = fileNameSave
            with open(file=self.fileName, mode="wt", encoding="utf-8") as saveFile:
                saveFile.write(self._getTextBoxContent())

    def saveAction_Clicked(self):
        if self.fileName:
            with open(file=self.fileName, mode="wt", encoding="utf-8") as saveFile:
                saveFile.write(self._getTextBoxContent())
        else:
            self.saveAsAction_Clicked()

    def _saveOrExitwithoutSaving_checkbox(self):
        qm = QMessageBox
        ret = qm.question(
            self, '', "Do you want to save changes?", qm.Yes | qm.No)
        if ret == qm.Yes:
            self.saveAction_Clicked()
        else:
            # self.newFileAction_Clicked()
            self._newFileActionSupport()

    def _newFileCheckSavedOrNot(self):
        if self.fileName:
            with open(file=self.fileName, mode="rt", encoding="utf-8") as fileOpen:
                fileOpenContent = fileOpen.read()
            if self._getTextBoxContent() != fileOpenContent:
                self._saveOrExitwithoutSaving_checkbox()
        else:
            self._saveOrExitwithoutSaving_checkbox()

    def _newFileActionSupport(self):
        try:
            self.saveAsAction_Clicked()
            if self.fileName:
                with open(file=self.fileName, mode="rt", encoding="utf-8") as fileOpen:
                    fileOpenContent = fileOpen.read()
                    self.textEdit.setPlainText(fileOpenContent)
        except Exception as e:
            print("Something is really wrong lol:", e)

    def newFileAction_Clicked(self):
        if self._getTextBoxContent():
            self._newFileCheckSavedOrNot()
        else:
            self._newFileActionSupport()

    def _printPreviewActionSupport(self, printer):
        self.textEdit.print_(printer)

    def printPreviewAction_Clicked(self):
        printer = QPrinter(QPrinter.HighResolution)
        previewDialog = QPrintPreviewDialog(printer, self)
        previewDialog.paintRequested.connect(self._printPreviewActionSupport)
        previewDialog.exec_()

    def printAction_Clicked(self):
        printer = QPrinter(QPrinter.HighResolution)
        dialog = QPrintDialog(printer, self)
        if dialog.exec_() == QPrintDialog.Accepted:
            self.textEdit.print_(printer)

    def undoAction_Clicked(self):
        # https://doc.qt.io/archives/qtforpython-5.12/PySide2/QtWidgets/QTextEdit.html#PySide2.QtWidgets.PySide2.QtWidgets.QTextEdit.setUndoRedoEnabled
        self.textEdit.undo()

    def redoAction_Clicked(self):
        self.textEdit.redo()

    def cutAction_Clicked(self):
        self.textEdit.cut()

    def copyAction_Clicked(self):
        self.textEdit.copy()

    def pasteAction_Clicked(self):
        self.textEdit.paste()

    def deleteAction_Clicked(self):
        # oldCopied = clipboard.paste()
        # self.cutAction_Clicked()
        # clipboard.copy(str(oldCopied))
        pass

    def findAction_Clicked(self):
        # flag = QTextDocument.FindBackward
        # self.textEdit.find('A', flag)
        pass

    def findNextAction_Clicked(self):
        pass

    def replaceAction_Clicked(self):
        pass

    def goToAction_Clicked(self):
        pass

    def selectAllAction_Clicked(self):
        self.textEdit.selectAll()

    def timeDateAction_Clicked(self):
        timeNow = datetime.now()
        self.textEdit.insertPlainText(
            str(timeNow.strftime("%H:%M - %d/%m/%Y")))

    def fontAction_Clicked(self):
        # https://doc.qt.io/qt-5/qfontdialog.html
        # https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QFontDialog.html
        # https://doc.qt.io/qt-5/qfontcombobox.html
        # https://www.tutorialspoint.com/pyqt/pyqt_qfontdialog_widget.htm
        font, ok = QFontDialog.getFont()

        # The original docs say that `font` is PySide2.QtGui.QFont and `ok` is bool
        # But when i tried it, the results were inverted
        try:
            self.textEdit.setFont(font)
        except TypeError:
            self.textEdit.setFont(ok)

    def zoomInAction_Clicked(self):
        self.textEdit.zoomIn(1)
        self.currentZoom += 1

    def zoomOutAction_Clicked(self):
        self.textEdit.zoomOut(1)
        self.currentZoom -= 1

    def restoreDefaultZoomAction_Clicked(self):
        self.textEdit.zoomIn(self.currentZoom)

    def viewHelpAction_Clicked(self):
        webbrowser.open(self.HELP_LINK)

    def sendFeedbackAction_Clicked(self):
        QMessageBox.about(self,
                          "Send Feedback",
                          "Email: zesta5j7k@gmail.com\nDiscord: ZeaCeR#5641")

    def aboutAction_Clicked(self):
        QMessageBox.about(self,
                          f"About {self.NAME}",
                          "Made by github user @hirusha-adi using the MIT License")

    def exit_app(self):
        self.close()


if __name__ == "__main__":
    myapp = QApplication(sys.argv)
    window = Window()
    myapp.exec_()
    sys.exit()
